internal/transport/http2_client.go

```diff
@@ -1282,96 +1283,73 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
-			// or come up with a way to just propagate the content-subtype if it was set?
-			// ie {"content-type": "application/grpc+proto"} or {"content-subtype": "proto"}
-			// in the metadata?
 			mdata[hf.Name] = append(mdata[hf.Name], hf.Value)
 			isGRPC = true
 		case "grpc-encoding":
 			s.recvCompress = hf.Value
 		case "grpc-status":
-			code, err := strconv.Atoi(hf.Value)
-			if err != nil {
-				grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed grpc-status: %v", err))
-				break
-			}
-			rawStatusCode = codes.Code(uint32(code))
+			rawStatus = hf.Value
 		case "grpc-message":
 			grpcMessage = decodeGrpcMessage(hf.Value)
 		case "grpc-status-details-bin":
-			v, err := decodeBinHeader(hf.Value)
+			sg, err := decodeGRPCStatusDetails(hf.Value)
 			if err != nil {
-				grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed grpc-status-details-bin: %v", err))
-				break
-			}
-			st := &spb.Status{}
-			if err := proto.Unmarshal(v, st); err != nil {
-				grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed grpc-status-details-bin: %v", err))
+				grpcStatus = fmt.Sprintf("transport: malformed grpc-status-details-bin: %v", err)
 				break
 			}
-			statusGen = status.FromProto(st)
+			statusGen = sg
 		case ":status":
-			// we only process :status if this is not grpc so no need to ALWAYS convert the value
-			mdata[hf.Name] = append(mdata[hf.Name], hf.Value)
-		case "grpc-tags-bin":
-			v, err := decodeBinHeader(hf.Value)
-			if err != nil {
-				grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed grpc-tags-bin: %v", err))
-				break
-			}
-			mdata[hf.Name] = append(mdata[hf.Name], string(v))
-		case "grpc-trace-bin":
-			v, err := decodeBinHeader(hf.Value)
-			if err != nil {
-				grpcStatus = status.New(codes.Internal, fmt.Sprintf("transport: malformed grpc-trace-bin: %v", err))
-				break
-			}
-			mdata[hf.Name] = append(mdata[hf.Name], string(v))
+			httpStatus = hf.Value
 		default:
 			if isReservedHeader(hf.Name) && !isWhitelistedHeader(hf.Name) {
 				break
 			}
 			v, err := decodeMetadataHeader(hf.Name, hf.Value)
 			if err != nil {
-				if logger.V(logLevel) {
-					logger.Errorf("Failed to decode metadata header (%q, %q): %v", hf.Name, hf.Value, err)
-				}
+				grpcStatus = fmt.Sprintf("transport: malformed %s: %v", hf.Name, err)
+				logger.Warningf("Failed to decode metadata header (%q, %q): %v", hf.Name, hf.Value, err)
 				break
 			}
 			mdata[hf.Name] = append(mdata[hf.Name], v)
 		}
 
 		// Account for the fact that we can set isGRPC at any time - it's possible
 		// that we'll set isGRPC AFTER an error has occurred.
-		if isGRPC && grpcStatus != nil {
-			t.closeStream(s, grpcStatus.Err(), true, http2.ErrCodeProtocol, grpcStatus, nil, endStream)
+		if isGRPC && grpcStatus != "" {
+			se := status.New(codes.Internal, grpcStatus)
+			t.closeStream(s, se.Err(), true, http2.ErrCodeProtocol, se, nil, endStream)
 			return
 		}
```

Comment:
Should this be moved outside the loop?

If so, then we can remove the `if isGRPC` part of it by putting it after the `if !isGRPC` block.